src/policy/marksweepspace/native_ms/global.rs

```diff
@@ -340,42 +369,42 @@ impl<VM: VMBinding> MarkSweepSpace<VM> {
         }
     }
 
-    pub fn generate_sweep_tasks(&self) -> Vec<Box<dyn GCWork<VM>>> {
-        // # Safety: ImmixSpace reference is always valid within this collection cycle.
-        let space = unsafe { &*(self as *const Self) };
-        self.chunk_map
-            .generate_tasks(|chunk| Box::new(SweepChunk { space, chunk }))
+    pub fn get_abandoned_block_lists(&self) -> &Mutex<AbandonedBlockLists> {
+        &self.abandoned
+    }
+
+    pub fn get_abandoned_block_lists_in_gc(&self) -> &Mutex<AbandonedBlockLists> {
+        &self.abandoned_in_gc
     }
 }
 
 use crate::scheduler::GCWork;
 use crate::MMTK;
 
-/// Chunk sweeping work packet.
-struct SweepChunk<VM: VMBinding> {
+struct PrepareChunkMap<VM: VMBinding> {
     space: &'static MarkSweepSpace<VM>,
     chunk: Chunk,
 }
 
-impl<VM: VMBinding> GCWork<VM> for SweepChunk<VM> {
+impl<VM: VMBinding> GCWork<VM> for PrepareChunkMap<VM> {
     fn do_work(&mut self, _worker: &mut GCWorker<VM>, _mmtk: &'static MMTK<VM>) {
         debug_assert!(self.space.chunk_map.get(self.chunk) == ChunkState::Allocated);
         // number of allocated blocks.
-        let mut allocated_blocks = 0;
-        // Iterate over all allocated blocks in this chunk.
-        for block in self
+        let occupied_blocks = self
             .chunk
             .iter_region::<Block>()
             .filter(|block| block.get_state() != BlockState::Unallocated)
-        {
-            if !block.attempt_release(self.space) {
-                // Block is live. Increment the allocated block count.
-                allocated_blocks += 1;
-            }
-        }
-        // Set this chunk as free if there is not live blocks.
-        if allocated_blocks == 0 {
+            .count();
+        if occupied_blocks == 0 {
+            // Set this chunk as free if there is no live blocks.
             self.space.chunk_map.set(self.chunk, ChunkState::Free)
+        } else {
+            // Otherwise this chunk is occupied, and we reset the mark bit if it is on the side.
+            if let MetadataSpec::OnSide(side) = *VM::VMObjectModel::LOCAL_MARK_BIT_SPEC {
+                for chunk in self.space.chunk_map.all_chunks() {
+                    side.bzero_metadata(chunk.start(), Chunk::BYTES);
```

Comment:
I found a strange thing.  We clear the mark bits, but we never clear block states before GC.  That means once a block becomes `BlockState::Marked`, it remains `BlockState::Marked` forever.

Reply:
Good observation. If what you said is true, it basically means we never release any block if it is ever marked. This could explain https://github.com/mmtk/mmtk-core/issues/688. I added code in this work packet to reset the block mark, and I will look further into this.

Reply:
I pushed two changes. https://github.com/mmtk/mmtk-core/pull/1112/commits/19e87aa72cf0e8a2ee2bcf2d26f09fa693961912 resets the block state in prepare. https://github.com/mmtk/mmtk-core/pull/1112/commits/a920da3892c7c9b81189aee7a3ae0a6b7b7a4b71 also releases unmarked block for lazy sweeping (otherwise dacapo tests still run OOM).

The min heap results after these fixes are here: https://github.com/mmtk/mmtk-core/issues/688#issuecomment-2058193888. They are improved a lot.